### PR TITLE
planner: switch to `AccessPath.PartialAlternativeIndexPaths` for IndexMerge OR path with MV index

### PR DESCRIPTION
**Pull request by time-and-fate**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/58361

### What changed and how does it work?

As the title says, the code path that generates IndexMerge OR path for MV indexes now generates `AccessPath.PartialAlternativeIndexPaths` instead of `AccessPath.PartialIndexPaths`.

Specifically:
- In `pkg/planner/core/indexmerge_unfinished_path.go`:
  - Now we store all alternatives instead of deciding on the best one.
  - Similarly, the `needSelection` checking related information is stored in `AccessPath.KeepIndexMergeORSourceFilter` now instead of being decided here.
  - Besides, a unified util function for estimation is extracted.
- In `pkg/planner/core/find_best_task.go`:
  - There's already logic that chooses the best one from the alternatives so we don't need implement it again.
  - The MV index should not be affected by the "avoid choosing the same index for partial paths" logic, so some extra code is still needed.
  - Use the `AccessPath.KeepIndexMergeORSourceFilter` information from the last step.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/planner/core/indexmerge_unfinished_path.go [hunk 1]
```diff
@@ -368,39 +366,41 @@ func buildIntoAccessPath(
...
-			alternativesForORBranch, func(a, b alternative) int {
-				return cmpAlternativesByRowCount(a.paths, b.paths)
-			})
-		if len(bestAlternative.paths) == 0 {
-			return nil
-		}
-		// Succeeded to get valid path(s) for this partial path.
-		partialPaths = append(partialPaths, bestAlternative.paths...)
-		needSelectionGlobal = needSelectionGlobal || bestAlternative.needSelection
+	// 2. TODO
```
Comment (Rustin170506) [inline pkg/planner/core/indexmerge_unfinished_path.go:385]: It seems you have already handled the table filters below. So what is this TODO for?

**Review by Rustin170506 — COMMENTED** [review]
Thanks!

File: pkg/planner/core/indexmerge_unfinished_path.go (hunk 1 above)
Comment (time-and-fate) [reply pkg/planner/core/indexmerge_unfinished_path.go:385]: Oh, I should add this in the next PR.
Some logic of other code paths will be moved here.

**Review by Rustin170506 — APPROVED** [review]
Thanks! :shipit:

File: pkg/planner/core/indexmerge_unfinished_path.go [hunk 2]
```diff
@@ -368,39 +366,41 @@ func buildIntoAccessPath(
...
-	if !needSelectionGlobal {
-		tableFilter = slices.Delete(tableFilter, orListIdxInAllConds, orListIdxInAllConds+1)
+	// For estimation, we need the decided partial paths. So we use a simple heuristic to choose the partial paths by
+	// comparing the row count just for estimation here.
+	pathsForEstimate := make([]*util.AccessPath, 0, len(allAlternativePaths))
+	for _, oneORBranch := range allAlternativePaths {
+		pathsWithMinRowCount := slices.MinFunc(oneORBranch, cmpAlternativesByRowCount)
+		pathsForEstimate = append(pathsForEstimate, pathsWithMinRowCount...)
 	}
+	possiblePath.CountAfterAccess = estimateCountAfterAccessForIndexMergeOR(ds, pathsForEstimate)
```
Comment (AilinKid) [inline pkg/planner/core/indexmerge_unfinished_path.go:401]: from the theoretical, we don't need to compute the count after acess at this point cuz it's still undetermined. but its seems to be used to update ds‘s stats introduced by @jiangtao
image
by now, it's fine, but not that elegant

**Review by AilinKid — APPROVED** [review]
LGTM

File: pkg/planner/core/indexmerge_unfinished_path.go (hunk 2 above)
Comment (time-and-fate) [reply pkg/planner/core/indexmerge_unfinished_path.go:401]: Yes. Besides this, the calculation of `DataSource.AccessPathMinSelectivity` also needs this field to be correctly set.

Comment (AilinKid) [reply pkg/planner/core/indexmerge_unfinished_path.go:401]: make sense

**Pull request merged**